common/djangoapps/third_party_auth/api/permissions.py

```diff
@@ -2,32 +2,91 @@
+    IsSuperuser,
+    JwtHasScope,
+    JwtRestrictedApplication,
+    NotJwtRestrictedApplication
+)
+from rest_condition import C
+from rest_framework.permissions import BasePermission, IsAuthenticated
 from third_party_auth.models import ProviderApiPermissions
 
+from openedx.core.lib.api.permissions import ApiKeyHeaderPermission
+
+log = logging.getLogger(__name__)
+
 
-class ThirdPartyAuthProviderApiPermission(permissions.BasePermission):
+class ThirdPartyAuthProviderApiPermission(BasePermission):
     """
     Allow someone to access the view if they have valid OAuth client credential.
-    """
-    def __init__(self, provider_id):
-        """ Initialize the class with a provider_id """
-        self.provider_id = provider_id
 
+    Deprecated: Only works for DOP oauth applications. To be removed as part of DOPrecation.
+
+    """
     def has_permission(self, request, view):
         """
         Check if the OAuth client associated with auth token in current request has permission to access
         the information for provider
         """
-        if not request.auth or not self.provider_id:
+        provider_id = view.kwargs.get('provider_id')
+        if not request.auth or not provider_id:
             # doesn't have access token or no provider_id specified
             return False
 
         try:
-            ProviderApiPermissions.objects.get(client__pk=request.auth.client_id, provider_id=self.provider_id)
+            ProviderApiPermissions.objects.get(client__pk=request.auth.client_id, provider_id=provider_id)
         except ProviderApiPermissions.DoesNotExist:
             return False
 
+        set_custom_metric('deprecated_ThirdPartyAuthProviderApiPermission', True)
         return True
+
+
+class JwtHasTpaProviderFilterForRequestedProvider(BasePermission):
+    """
+    The JWT used to authenticate contains the appropriate tpa_provider
+    filter for the requested provider.
+    """
+    message = 'JWT missing required tpa_provider filter.'
+
+    def has_permission(self, request, view):
+        """
+        Ensure that the provider_id kwarg provided to the view exists exists
+        in the tpa_provider filters in the JWT used to authenticate.
+        """
+        provider_id = view.kwargs.get('provider_id')
+        if not provider_id:
+            log.warning("Permission JwtHasTpaProviderFilterForRequestedProvider requires a view with provider_id.")
+            return False
+
+        jwt_filters = decode_jwt_filters(request.auth)
+        for filter_type, filter_value in jwt_filters:
+            if filter_type == 'tpa_provider' and filter_value == provider_id:
+                return True
+
+        log.warning(
+            "Permission JwtHasTpaProviderFilterForRequestedProvider: required filter tpa_provider:%s was not found.",
+            provider_id,
+        )
+        return False
+
+
+# TODO: Remove ApiKeyHeaderPermission. Check deprecated_api_key_header custom metric for active usage.
```

Comment:
**Note:** In Prod, I see lots of calls in NewRelic that look unauthenticated, which I am guessing might be using `ApiKeyHeaderPermission`. Not sure when and if we would want to migrate these callers to use JWTs with scopes?

See: `SELECT * from Transaction where name like '%UserMappingView%' and response.status = '200'`